### ISSUE TITLE
M3U files required by cd-based beetle cores but not supported well at all by UI and features available for CUE scans

### ISSUE THREAD

**Issue by birdybro** [libretro/RetroArch#11352]
## Description

Multiple problems with M3U files in the user interface as of version 19.0 (and as long as beetle cores have required m3u for multi-disc on-the-fly changing). M3U files are a requirement for using beetle cores that are for CD games with multi-disc support. Here's the problems that need to be addressed as it's very unintuitive to use M3U's but still required:

* Scanning directory by default does not scan in .m3u files, this has to be done manually. This doesn't make sense if it's required to use them for basic functionality.
* .m3u files are not checked for in the game information database (e.g. Chrono cross cue files get thumbnails and database information, m3u gets no thumbnails or information, just the title)
* If a game has an .m3u file being utilized (meaning it's a multi-disc game since this is necessary), the scanner should ignore the cue files of the same name/database entry, but currently only gets cue files, and as said before ignores .m3u. It ought to include only .m3u files for games that have them, and cue files for games that don't have an .m3u, to not import lots of unnecessary games. manual scanning does not help with this problem as you will get either only .m3u (which most games don't require), or both cue and m3u, or you exclude m3u as your only 3 options. That or you manually edit the playlist file laboriously if you have a large collection.

### Expected behavior

I expect to "scan directory" the PSX directory and have my .m3u supported games have a thumbnail, game information from teh database, and my cue-only games to be scanned in like before, with the cue's for m3u-supported games being excluded from the list.

### Actual behavior

Scan directory adds only cue files from the psx directory, ignores m3u, and if you manual scan the directory for .m3u, then it still doesn't use the games database and therefore gets no game information or thumbnails for the m3u entries. If you scan manually for m3u only, you are excluding cue, if you scan manually for cue, you miss m3u, and if you scan manually for both, you have excess cue entries from .m3u enabled games.

### Steps to reproduce the bug

Import --> scan directory --> a psx folder with a mixed set of m3u-using games and cue-using games --> look at the psx playlist and see no m3u files --> manual scan for m3u --> update thumbnails --> m3u games are now added, but no game information and no thumbnails and redundant cue files for each disc have to be manually removed

### Version/Commit
You can find this information under Information/System Information

- RetroArch: v1.9.0

### Environment information

- OS: Windows 10 Home Feature Update 2004

**Comment by i30817** [comment]
1. It actually can be done automatically, the RA project is just resistant to doing so. For instance i implemented [this for linux a while ago](https://gist.github.com/i30817/ba37fbb2b3c6e34ff926ad833f465055).
2. database problem. There is a technical solution but i requires discipline from contributors, so it wasn't done, badum tish.
3.  The script above solved it for me.

**Comment by birdybro** [comment]
1. You might be confusing this with the previous issue I raised with an automatic m3u generator feature request, which we both talked about back then too. :) - This is in regards to the scan directory behavior not being appropriate for the most used core when it comes to multi-disc games.

2. How much technical knowledge is it to do this? Is it simple data entry for the most part? If it is, then I guess I could start knocking them out.

3. This doesn't generate a playlist (.lpl) file though. I remember when we chatted about this last in my feature request here previously that the idea was to integrate it.

My issue raised here assumes that the user has already made their own .m3u files, and the scanning of the directory doesn't work as it ought to. I think I will spin off your gist like i had previously planned to since I have gotten slightly more experience messing with code since we last chatted about it, just as a side thing.

**Comment by i30817** [comment]
Guess i forgot. About 2, it's a scripting problem and a 'enforce contributors contribute files in a specific way' problem.

The idea is simple. Someone contributes a image with listing the game console that applies, and the file(s) with a filename like Xenogears (USA).png - xenogears the game has two discs, but assume the image is the same for now. If they want a particular edition or cd to be different they can be more specific, aka 2 commits, 'Xenogears (USA) (Disk 1).png' and 'Xenogears (USA) (Disk 2).png'


Now for the script. Periodically the person updating the thumnail database runs a script that:
1. takes the console name in the commit (easily done by seeing the parent dir in the libretro database scheme)
2. searches for game filenames that match the part of the name without parenthesis or extension, and normalized spaces (this means either 1 space for the 'gaps' you remove or none if the game was a the end before the png extension. Why remove the groups? Because some dumping groups have nonstandard ideas and can easily place extra groups in between what you expected.
3. Anyway, you're not done! Take the groups you took, and for two types in particular, multiply them. '(Rev. 1.1)' becomes both that and  '(1.1)', '(Disc 1)' becomes both that and '(Disc 1 of #game_discs)' - which you got from step 2 -, and '(Cd 1)' and '(Cd 1 of #)' - all the variations for this of the common dumping groups at least.
4. filter again with the all the groups created like this, and consider those 'multiplied' as a set where 'any' could match'.
5. You have listing of all the games that primordial filename applies to in the libretro database now and could make the pngs.
6. Remember back in step 1 when i said you could have 'more specific' names? Yeah, if the file already exists, don't replace it until the admin or a commit removes it.

**Comment by i30817** [comment]
edited the algorithm above to hopefully be a bit more clear. Anyway, i find it doable, certainly in python or whatever libretro database uses.

**Comment by i30817** [comment]
Btw @birdybro , i remember @jdgleaver implementing something like the filtering for files inside m3u scanned that you mention doesn't work? Am i just imagining it @jdgleaver?

**Comment by jdgleaver** [comment]
Yes, if you scan an M3U file with the manual scanner, it will automatically exclude any files that are referenced inside. Also, if you clean a playlist (`Manage Playlists > Clean Playlist`), it will remove any individually-listed items that are duplicated inside M3Us.

**Comment by birdybro** [comment]
I must have been doing something wrong with the manual scanner, I have confirmed that it works, but that doesn't address the other issues. Can't this same behavior from the manual scanner be imported into the automatic scanner?

**Comment by birdybro** [comment]
As far as .m3u generation, I found a great tool for this which I believe works cross-platform, maybe the method could be implemented.

https://github.com/tralph3/Cue-Maker

Or maybe this person would be interested in contributing to retroarch with the m3u generation python script they made @tralph3

**Comment by tralph3** [comment]
I actually made the script thinking about retroarch, but I don't think retroarch should be in charge of making m3u or cue files. I do want it to detect the m3u automatically and ignore the cue files inside it so I have a single entry with multi disc games (this is currently possible, but I have to manually remove the cues and add the m3u with the desktop app).
